### ISSUE TITLE
Problem: Undocumented breaking change in Galaxy role installation from v2.10.31

### ISSUE THREAD

**Issue by ShaneMcC** [ansible-semaphore/semaphore#2519]
### Issue

Commit c76f04a39b70decfa007787a085bc9b7391cf752 makes a breaking change in how semaphore installs roles/collections.

Previously, semaphore would install roles from:

- `<playbook_dir>/collections/requirements.yml` or `<repo_dir>/collections/requirements.yml`
- `<playbook_dir>/roles/requirements.yml` or `<repo_dir>/roles/requirements.yml`

But this change now makes it only install relative to the playbook path:

- `<playbook_dir>/collections/requirements.yml` or `<playbook_dir>/requirements.yml`
- `<playbook_dir>/roles/requirements.yml` or `<playbook_dir>/requirements.yml`

Ignoring `<repo_dir>/roles/requirements.yml` and `<repo_dir>/collections/requirements.yml` files entirely that were previously used (similar to anisble tower which looks for these at the top-level)

In cases where the playbook is in the repo root, this change makes no difference, however in cases where we are running a playbook from a subdirectory this causes problems.

This was documented in the release for `v2.10.31` as `Handle <playbook_dir>/requirements.txt by ansible-galaxy.` but no mention was made around this removing the previous behaviour.

If we want to also look at `<playbook_dir>/requirements.yml` first then this should be in addition to the previous behaviour not instead of.

### Impact

Ansible (task execution)

### Installation method

Docker

### Database

Postgres

### Browser

Chrome

### Semaphore Version

v2.10.35-a71a3c5-1730103926

**Comment by jp1337** [comment]
Hi @ShaneMcC and @fiftin!
In addition to the original problem, we encountered another one today.

If you merge roles and collections together in a requirements.yml file and place it in the playbook directory, no roles are installed.

The following test case:
1. We had placed a requirements.yml with only one role directly next to the playbook in its directory.
2. When the task was executed, an attempt was first made to install the collections from the requirements.yml.
3. This created a file with the md5 hash of the requirements.yml.
4. During the role installation, the requirements.yml was then skipped because the md5 hash had not changed.

So currently the only way is via a roles and a collections subfolder in the playbook directory. From our point of view, no other way is possible.

Example structure:
```sh
playbooks/subdir/
├── collections
│   └── requirements.yml
├── playbook.yml
├── roles
     └── requirements.yml
```

In addition to this structure the logic should also work for the following example where roles and requirements are places in a single requirements.yml file.

This structure is officially supported and documented by ansible in this guide:
[Installing roles and collections from the same requirements.yml file](https://docs.ansible.com/ansible/latest/galaxy/user_guide.html#installing-roles-and-collections-from-the-same-requirements-yml-file)

```sh
playbooks/subdir/
├── playbook.yml
└── requirements.yml
```

Since we have a large number of playbooks and also a large number of folders, we had to help ourselves out with a Python script. This automatically generates the requirements.yml files and the corresponding collections and roles folders.
This script would not be needed if we could use a single requirements.yml file for roles and requirements like described in the issue.

**Comment by ShaneMcC** [comment]
@jp1337 I think that is a separate issue than this one - semaphore has always needed separate `requirements.yml` files for collections and roles. But previously it always at least read them at the top-level not only within the playbook directory.

(I would also prefer that it was able to use a single file for both, but it has always been separate as long as I've used it and it at least worked as-expected before once I separated them)

**Comment by svennd** [comment]
I agree with the @ShaneMcC , I suggested a change to support both. For now I used a ansible.cfg to a absolute local path where the collections are installed for me. But ideally it follows the requirements inside the repo. 

see https://github.com/semaphoreui/semaphore/pull/661#issuecomment-821142548 for explanation how to use ansible.cfg;

**Comment by xenion1987** [comment]
I still don't get it how to install roles correctly.

```
.
├── ansible.cfg
├── group_vars
│   └── all.yml
├── host_vars
│   └── localhost.yml
├── play_init.yml
├── README.md
├── requirements.txt
├── requirements.yml
└── roles
    └── requirements.yml
```

`ansible.cfg`
```
[defaults]
# Path configurations
collections_path = ./
roles_path = ./roles
```

`./requirements.yml`
```yml
---
collections:
  - ansible.posix
  - community.general
```

`./roles/requirements.yml`
```yml
---
roles:
  - name: user_management
    src: Xenion1987.user_management
  - name: sshd
    src: Xenion1987.sshd
  - name: sudoers
    src: Xenion1987.sudoers
```


![Image](https://github.com/user-attachments/assets/63bf853d-c6f9-4525-879c-446606bee676)

**Comment by svennd** [comment]
@Xenion1987 
I solved it by creating (and maintaining) them outside of the semaphore : 
`collections_path = /home/svennd/collections:./collections`

Since I only needed a few that worked but its not ideal.

**Comment by xenion1987** [comment]
Thank you for sharing your workaround, @svennd.

Now, I've set it up this way:

```
~/Docker/semaphore# chown -R 1001:0 config/etc/ansible
~/Docker/semaphore# tree -aFL 5
./
├── config/
│   └── etc/
│       └── ansible/
│           ├── ansible.cfg
│           ├── collections/
│           │   └── ansible_collections/
│           ├── requirements.txt
│           ├── requirements.yml
│           └── roles/
│               ├── sshd/
│               ├── sudoers/
│               └── user_management/
├── docker-compose.yml
└── .env
```

`docker-compose.yml`

```yml
---
services:
  semaphore:
    image: semaphoreui/semaphore:latest
    (...)
    volumes:
      - /etc/localtime:/etc/locatime:ro
      - ${PWD}/config/etc/ansible:/etc/ansible:rw # <<--- Mounting my stuff into the container
```

I install/update collections and roles inside the docker container, so I do not need to install `ansible-galaxy` on the host itself:

```sh
docker compose exec -it semaphore \
  ansible-galaxy install -r /etc/ansible/requirements.yml
```

Based on the `ansible.cfg`, collections and roles are getting installed in `/etc/ansible/`:

`~/Docker/semaphore# docker compose exec -it semaphore cat /etc/ansible/ansible.cfg`

```ini
[defaults]
(...)
# Path configurations
collections_path = /etc/ansible/collections
roles_path = /etc/ansible/roles
(...)
```

`config/etc/ansible/requirements.yml`

```yml
---
collections:
  - ansible.posix
  - community.general
roles:
  - name: user_management
    src: Xenion1987.user_management
  - name: sshd
    src: Xenion1987.sshd
  - name: sudoers
    src: Xenion1987.sudoers
```

**Comment by alexbarton** [comment]
This affects us as well and breaks all workflows … :-(

Any chance to get the old behaviour back, looking for `requirements.yml` in `collections/` and `roles/` sub-folders?

Or even better, supporting a `requirements.yml` file at the project (=repository) root?

**Comment by alexbarton** [comment]
BTW: I guess this relates to (duplicates?) #1609?

**Comment by alexbarton** [comment]
Looking for requirements.yml in the project root is fixed in https://github.com/semaphoreui/semaphore/pull/2687 and works for me.

**Comment by fiftin** [comment]
Fixed. Thank you :)
